src/io/fasta.rs

```diff
@@ -252,12 +361,73 @@ struct IndexRecord {
 
 
 /// A sequence record returned by the FASTA index.
+#[derive(Debug, PartialEq)]
 pub struct Sequence {
     pub name: String,
     pub len: u64,
 }
 
 
+pub struct IndexedReaderIterator<'a, R: io::Read + io::Seek + 'a> {
+    reader: &'a mut IndexedReader<R>,
+    record: IndexRecord,
+    bases_left: u64,
+    line_offset: u64,
+    buf: Vec<u8>,
+    buf_idx: usize,
+    buf_len: usize,
+}
+
+
+impl<'a, R: io::Read + io::Seek + 'a> IndexedReaderIterator<'a, R> {
+    fn fill_buffer(&mut self) -> io::Result<()> {
+        assert!(self.bases_left > 0);
+
+        while self.buf_idx >= self.buf_len {
+            let bases_read = self.reader
+                .read_line(&self.record,
+                           &mut self.line_offset,
+                           self.bases_left,
+                           &mut self.buf)?;
+
+            self.buf_idx = 0;
+            self.buf_len = bases_read as usize;
+            self.bases_left -= bases_read;
+        }
+
+        Ok(())
+    }
+}
+
+
+impl<'a, R: io::Read + io::Seek + 'a> Iterator for IndexedReaderIterator<'a, R> {
+    type Item = io::Result<u8>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.buf_idx < self.buf_len {
+            let item = Some(Ok(self.buf[self.buf_idx]));
```

Comment:
It is true that these iterators sacrifice speed for convenience. What you get is the utility of iterators, and a controlled memory usage without having manage a buffer (often multiple buffers, in my case) and without having to perform a seek between every chunk you read. I tend to prefer simpler code using iterators, unless profiling showed that a significant percentage of the runtime was wasted by doing so, but that is unlikely for all but the simplest analyses.

The number of copies is also the same as the existing implementation of 'IndexedReader::read', which first reads into 'buf' and then copies the bases into 'seq'. I briefly toyed with an implementation of 'IndexedReader::read' that only used a temporary buffer to dump whitespace, in order to reduce copying, but I do not recall seeing any improvement over the alternative. If you are interested, then I can try to redo that implementation and do a benchmark.

I can probably also improve IndexedReader::seek_to to avoid unnecessary seeks independant of whether iterators are added, which would help performance when doing block reads. That would improve the case where that overhead in unacceptable, but you still want to control memory usage.